### DISCUSSION TITLE
Traits, skills, and skill trees

### DISCUSSION THREAD

**Discussion by InspectorCaracal** [evennia/evennia#2685] [General Discussion]
So I've started trying to build the skill system for my game and I thought it would work well to use the existing Traits contrib for it, since it's designed specifically for that kind of thing. The trouble is, the way I want to structure my skills involves a sort of "skill inheritance" which, as far as I can tell, the Traits system doesn't support and I can't figure out how to expand it to do so in a way that isn't terrible.

The short version of the idea is that you have general skills, e.g. sewing, and then specializations of that skill, e.g. tailoring or quilting, and specialization checks include a modifier from the parent skill.

Going into a bit more detail: Sewing by itself is a skill, and general sewing tasks would reference your sewing skill. Trying to do a more specialized kind of sewing, however, such as making a quilt, would reference the specialization skill. Improving your basic sewing skill should make you better at making quilts, but not as good as you can be if you improve your quilting - and improving quilting won't make you any better at tailoring.

On a technical level, the idea is that when doing something which uses a specialization skill, it has its own skill level, but it _also_ inherits a fraction of the parent skill - let's say half, for ease of example. So if you have a sewing level of 10 and have no quilting-specific skill, your quilting skill check would be 5. But if you had a sewing level of 10 and a quilting-specific level of 2, your quilting check would be 7. (Ideally I could nest skills as deeply as I need to, with each successive step down making the root skill have less of an impact on the specialization as the fractions compound.)

Unfortunately for me, with the Traits module, there doesn't seem to be any straightforward way to make trait values rely on or cross-reference other traits, so I suspect it won't work out for me to use it.

Has anyone tried doing something like this before? What approach worked or didn't work for you?

**Comment by strikaco** [comment]
Here's a quick-and-dirty skill tree implementation you can use with your existing TraitsHandler.

```
# These classes represent the TraitsHandler and a typeclassed Object.
# Just ignore this stuff; necessary setup for demonstration
class DemoSkills(object):
    nature = 0.5
    hunting = 1
    swimming = 2
    diving = 3

class DemoObject(object):
    traits = DemoSkills()

# ---
# Here's where you come in. Copy/tweak as needed
SKILLS = {}

class Skill(object):

    def __hash__(self):
        return hash(self.key)

    def __init__(self, key, parent=None):
        self.key = key
        self.parent = parent
        SKILLS[self.key] = self

    def get(self, obj):
        # Redefine this to get the value from wherever (traitshandler, ndb, etc.)
        value = getattr(obj, self.key, 0)

        if not self.parent:
            return value

        # Redefine this to account for amortization or whatever
        return self.parent.get(obj) + value

# Define skills hierarchy
# All of these can be iterated via SKILLS.items()
NATURE = Skill('nature')
HUNTING = Skill('hunting', NATURE)

SWIMMING = Skill('swimming', NATURE)
DIVING = Skill('diving', SWIMMING)

# ---
# POC
char = DemoObject()
# Show all AGGREGATE skills (for stats screen or whatever)
for k,v in SKILLS.items():
    print(k, SKILLS[k].get(char.traits))
print()

# Calculate the aggregate Diving skill. 
# Diving is a specialization of Swimming, which rolls up to Nature
agg_diving = DIVING.get(char.traits)
print('diving', agg_diving)
assert agg_diving == 5.5

```

**Reply by InspectorCaracal** [reply]
Hrm. I tried something like `self.parent = parent` as my first attempt and it didn't work at all, but it's possible I did something wrong at the time so I'll give it another shot.

**Reply by strikaco** [reply]
Trees are tricky; unfortunately there is no builtin data structure for it, just a lot of custom implementations that are easy to screw up.

Case in point, I think I did the math opposite of what you requested-- I give *additional* weight to subspecialties. `return self.parent.get(obj) + value` should probably be changed to a flat % decrease of the parent or something, otherwise changing it to a subtraction operation yields negative values.

(Try `return (self.parent.get(obj) * 0.25) + value`, which should be more the right direction.)

**Reply by InspectorCaracal** [reply]
Ah, well, the problem I had wasn't in structuring the dependency so much as not being able to access the data because the handler couldn't store/retrieve self-referential data in a trait e.g. a different trait, and I still haven't figured out a workaround that doesn't require rewriting the traits handler (or honestly, even one that _does_ require rewriting the traits handler).

**Comment by Griatch** [comment]
Did you try something like this? 

```python 
from evennia.contrib.rpgs.traits import CounterTrait

class DependencyTrait(CounterTrait):
    trait_type = "dependent"

    @property
    def value(self): 
         character = self.character
         dependency_name = self.dependency_name
         dependent_value = character.traits.get(dependency_name).value         
         return (dependent_value * 0.25) + self.value

# The TraitTailloring needs to be added in a module checked by settings.TRAIT_CLASS_PATHS

# there is no `.character` or `dependency_name` available on the Trait by default, 
# so we must initialize the trait this way; the char-reference should pickle properly

char.traits.add("sewing", trait_type="counter", base=10)
char.traits.add("quilting", trait_type="dependent", base=2, character=char, dependency_name="sewing")
```
